tinylog-api/src/main/java/org/tinylog/configuration/PropertiesConfigurationLoader.java

```diff
@@ -0,0 +1,132 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package org.tinylog.configuration;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Properties;
+import java.util.regex.Pattern;
+
+import org.tinylog.Level;
+import org.tinylog.provider.InternalLogger;
+import org.tinylog.runtime.RuntimeProvider;
+
+/**
+ * Standard internal property loader for tinylog.
+ *
+ * <p>
+ * By default, the configuration will be loaded from {@code tinylog.properties} in the default package. Another
+ * configuration file can be loaded by setting the system property {@code tinylog.configuration}. The configuration file
+ * can be a resource in the classpath, a file from file system or an URL
+ * </p>
+ *
+ * <p>
+ * Alternately configuration properties can be set via system properties. These properties must be prefixed by
+ * "{@code tinylog.}". For example: "{@code level = debug}" becomes "{@code tinylog.level=debug}". If a configuration
+ * property exists as system property and in configuration file, the system property will win.
+ * </p>
+ */
+public class PropertiesConfigurationLoader implements ConfigurationLoader {
+	private static final String[] CONFIGURATION_FILES = new String[] {
+		"tinylog-dev.properties",
+		"tinylog-test.properties",
+		"tinylog.properties"
+	};
+	
+
+	private static final String CONFIGURATION_PROPERTY = Configuration.PROPERTIES_PREFIX + "configuration";
+	private static final Pattern URL_DETECTION_PATTERN = Pattern.compile("^[a-zA-Z]{2,}:/.*");
+	
+	/** */
+	public PropertiesConfigurationLoader() {
+		
+	}
+	
+	/**
+	 * Loads all configuration properties.
+	 *
+	 * @return Found properties
+	 */
+	@Override
+	public Properties load() {
+		Properties properties = new Properties();
+
+		String file = System.getProperty(CONFIGURATION_PROPERTY);
+		InputStream stream = null;
+		try {
+			if (file != null) {
+				if (URL_DETECTION_PATTERN.matcher(file).matches()) {
+					stream = new URL(file).openStream();
+				} else {
+					stream = RuntimeProvider.getClassLoader().getResourceAsStream(file);
+					if (stream == null) {
+						stream = new FileInputStream(file);
+					}
+				}
+				load(properties, stream);
+			} else {
+				for (String configurationFile : getConfigurationFiles()) {
+					file = configurationFile;
+					stream = RuntimeProvider.getClassLoader().getResourceAsStream(file);
+					if (stream != null) {
+						load(properties, stream);
+						break;
+					}
+				}
+			}
+		} catch (IOException ex) {
+			InternalLogger.log(Level.ERROR, "Failed loading configuration from '" + file + "'");
+		} finally {
+			if (stream != null) {
+				try {
+					stream.close();
+				} catch (IOException ex) {
+					// Do nothing
+				}
+			}
+		}
+
+		Configuration.mergeSystemProperties(properties);
+		Configuration.resolveProperties(properties, EnvironmentVariableResolver.INSTANCE, SystemPropertyResolver.INSTANCE);
```

Comment:
I propose to call both methods in `Configuration.load()`. In this case, all future configuration loaders will benefit from merging and resolving properties automatically.

Reply:
Yes you are right. This makes sense.